File: integration_tests/java_client_tests_reactor_real/client-app/src/test/java/dev/graphlink/test/CacheTest.java
package dev.graphlink.test;

import dev.graphlink.test.generated.client.GraphLinkClient;
import dev.graphlink.test.generated.enums.UserStatus;
import dev.graphlink.test.generated.inputs.AddressInput;
import dev.graphlink.test.generated.inputs.CreateUserInput;
import dev.graphlink.test.generated.types.GetCachedUserResponse;
import dev.graphlink.test.generated.types.GetStaleUserResponse;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static dev.graphlink.test.Fixtures.*;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Cache is 100% client-side logic. These tests use MockAdapter so they can
 * count exact adapter calls and simulate network failure, independent of a
 * running server.
 */
class CacheTest {

    private MockAdapter adapter;
    private GraphLinkClient client;

    private static final AddressInput kAddressInput = AddressInput.builder()
            .street("123 Main St").city("Springfield").country("US").zip("12345").build();

    private static final CreateUserInput kCreateInput = CreateUserInput.builder()
            .name("Alice Smith").email("alice@test.com")
            .status(UserStatus.ACTIVE).address(kAddressInput).build();

    @BeforeEach
    void setUp() {
        adapter = new MockAdapter();
        client = newClient(adapter, new MockWebSocketAdapter());
    }

    // ── Cache hit ─────────────────────────────────────────────────────────────

    @Nested
    class CacheHit {

        @BeforeEach
        void register() {
            adapter.registerData("getCachedUser", Map.of("getCachedUser", kUserAliceJson));
        }

        @Test
        void firstCall_hitsAdapter() {
            client.queries.getCachedUser("user-1").block();
            assertEquals(1, adapter.getCallCount());
        }

        @Test
        void secondCallSameArg_servedFromCache_noExtraAdapterCall() {
            client.queries.getCachedUser("user-1").block();
            client.queries.getCachedUser("user-1").block();
            assertEquals(1, adapter.getCallCount());
        }

        @Test
        void cachedResponse_hasCorrectData() {
            client.queries.getCachedUser("user-1").block();
            GetCachedUserResponse res = client.queries.getCachedUser("user-1").block();
            assertNotNull(res.getGetCachedUser());
            assertEquals("user-1", res.getGetCachedUser().getId());
        }

        @Test
        void differentArg_bypassesCache() {
            adapter.registerData("getCachedUser", Map.of("getCachedUser", kUserBobJson));
            client.queries.getCachedUser("user-1").block();
            client.queries.getCachedUser("user-2").block();
            assertEquals(2, adapter.getCallCount());
        }
    }

    // ── Tag-based invalidation ────────────────────────────────────────────────

    @Nested
    class TagInvalidation {

        @Test
        void createCachedUser_invalidatesUsersTag_nextGetCachedUserRefetches() {
            adapter.registerData("getCachedUser", Map.of("getCachedUser", kUserAliceJson));
            adapter.registerData("createCachedUser", Map.of("createCachedUser", kUserAliceJson));

            client.queries.getCachedUser("user-1").block();
            assertEquals(1, adapter.getCallCount());

            client.mutations.createCachedUser(kCreateInput).block();
            assertEquals(2, adapter.getCallCount());

            client.queries.getCachedUser("user-1").block();
            assertEquals(3, adapter.getCallCount());
        }

        @Test
        void resetAll_invalidatesAllTags_allCachedQueriesRefetch() {
            adapter.registerData("getCachedUser", Map.of("getCachedUser", kUserAliceJson));
            adapter.registerData("getCachedPost", Map.of("getCachedPost", kPostJson));
            adapter.registerData("resetAll", Map.of("resetAll", true));

            client.queries.getCachedUser("user-1").block();
            client.queries.getCachedPost("post-1").block();
            assertEquals(2, adapter.getCallCount());

            client.mutations.resetAll().block();
            assertEquals(3, adapter.getCallCount());

            client.queries.getCachedUser("user-1").block();
            client.queries.getCachedPost("post-1").block();
            assertEquals(5, adapter.getCallCount());
        }
    }

    // ── Cross-tag isolation ───────────────────────────────────────────────────

    @Nested
    class CrossTagIsolation {

        @Test
        void usersTagInvalidation_doesNotAffect_postsCache() {
            adapter.registerData("getCachedPost", Map.of("getCachedPost", kPostJson));
            adapter.registerData("createCachedUser", Map.of("createCachedUser", kUserAliceJson));

            client.queries.getCachedPost("post-1").block();
            assertEquals(1, adapter.getCallCount());

            client.mutations.createCachedUser(kCreateInput).block();
            assertEquals(2, adapter.getCallCount());

            client.queries.getCachedPost("post-1").block();
            assertEquals(2, adapter.getCallCount());
        }

        @Test
        void transferPost_invalidatesBothTags() {
            adapter.registerData("getCachedUser", Map.of("getCachedUser", kUserAliceJson));
            adapter.registerData("getCachedPost", Map.of("getCachedPost", kPostJson));
            adapter.registerData("transferPost", Map.of("transferPost", kPostJson));

            client.queries.getCachedUser("user-1").block();
            client.queries.getCachedPost("post-1").block();
            assertEquals(2, adapter.getCallCount());

            client.mutations.transferPost("post-1", "user-2").block();
            assertEquals(3, adapter.getCallCount());

            client.queries.getCachedUser("user-1").block();
            client.queries.getCachedPost("post-1").block();
            assertEquals(5, adapter.getCallCount());
        }
    }

    // ── listCachedUsers ───────────────────────────────────────────────────────

    @Nested
    class ListCachedUsers {

        @Test
        void listCachedUsers_isCachedAfterFirstCall() {
            adapter.registerData("listCachedUsers",
                    Map.of("listCachedUsers", List.of(kUserAliceJson)));
            client.queries.listCachedUsers().block();
            client.queries.listCachedUsers().block();
            assertEquals(1, adapter.getCallCount());
        }

        @Test
        void createCachedUser_invalidatesListCachedUsers() {
            adapter.registerData("listCachedUsers",
                    Map.of("listCachedUsers", List.of(kUserAliceJson)));
            adapter.registerData("createCachedUser", Map.of("createCachedUser", kUserAliceJson));

            client.queries.listCachedUsers().block();
            client.mutations.createCachedUser(kCreateInput).block();
            client.queries.listCachedUsers().block();
            assertEquals(3, adapter.getCallCount());
        }
    }

    // ── staleIfOffline ────────────────────────────────────────────────────────

    @Nested
    class StaleIfOffline {

        @Test
        void staleData_isReturnedWhenNetworkFails_afterTtlExpires() throws InterruptedException {
            adapter.registerData("getStaleUser", map("getStaleUser", kUserAliceJson));

            client.queries.getStaleUser("user-1").block();
            assertEquals(1, adapter.getCallCount());

            Thread.sleep(1100);

            adapter.simulateFailure = true;
            GetStaleUserResponse res = client.queries.getStaleUser("user-1").block();

            assertNotNull(res.getGetStaleUser());
            assertEquals("user-1", res.getGetStaleUser().getId());
        }

        @Test
        void nonStaleEntry_throwsWhenNetworkFails() {
            adapter.registerData("getCachedUser", Map.of("getCachedUser", kUserAliceJson));
            adapter.simulateFailure = true;
            assertThrows(RuntimeException.class,
                    () -> client.queries.getCachedUser("user-1").block());
        }
    }
}

File: integration_tests/swift_client_tests_real/Tests/GraphLinkGeneratedTests/CacheTest.swift
import XCTest
@testable import GraphLinkGenerated

// Even in the real-server tier, caching call-counts can only be verified via a
// MockAdapter (there is no way to observe real-server round-trips otherwise) —
// mirrors the Kotlin CacheTest.kt in kotlin_client_tests_real, which does the same.

private let kAddressInput = AddressInput(street: "123 Main St", city: "Springfield", country: "US", zip: "12345")
private let kCreateInput = CreateUserInput(
    name: "Alice Smith",
    email: "alice@test.com",
    status: .active,
    address: kAddressInput
)

// ── Cache hit ────────────────────────────────────────────────────────────────

final class CacheTest_CacheHit: XCTestCase {
    var adapter: MockAdapter!
    var client: GraphLinkClient!

    override func setUp() {
        adapter = MockAdapter()
        client = Fixtures.newClient(adapter, MockWebSocketAdapter())
        adapter.registerData("getCachedUser", n(("getCachedUser", Fixtures.kUserAliceJson)))
    }

    func test_firstCall_hitsAdapter() async throws {
        _ = try await client.queries.getCachedUser(id: "user-1")
        XCTAssertEqual(adapter.getCallCount(), 1)
    }

    func test_secondCallSameArg_servedFromCache_noExtraAdapterCall() async throws {
        _ = try await client.queries.getCachedUser(id: "user-1")
        _ = try await client.queries.getCachedUser(id: "user-1")
        XCTAssertEqual(adapter.getCallCount(), 1)
    }

    func test_cachedResponse_hasCorrectData() async throws {
        _ = try await client.queries.getCachedUser(id: "user-1")
        let res = try await client.queries.getCachedUser(id: "user-1")
        XCTAssertNotNil(res.getCachedUser)
        XCTAssertEqual(res.getCachedUser.id, "user-1")
    }

    func test_differentArg_bypassesCache() async throws {
        adapter.registerData("getCachedUser", n(("getCachedUser", Fixtures.kUserBobJson)))
        _ = try await client.queries.getCachedUser(id: "user-1")
        _ = try await client.queries.getCachedUser(id: "user-2")
        XCTAssertEqual(adapter.getCallCount(), 2)
    }
}

// ── Tag-based invalidation ───────────────────────────────────────────────────

final class CacheTest_TagInvalidation: XCTestCase {
    var adapter: MockAdapter!
    var client: GraphLinkClient!

    override func setUp() {
        adapter = MockAdapter()
        client = Fixtures.newClient(adapter, MockWebSocketAdapter())
    }

    func test_createCachedUser_invalidatesUsersTag_nextGetCachedUserRefetches() async throws {
        adapter.registerData("getCachedUser", n(("getCachedUser", Fixtures.kUserAliceJson)))
        adapter.registerData("createCachedUser", n(("createCachedUser", Fixtures.kUserAliceJson)))

        _ = try await client.queries.getCachedUser(id: "user-1")
        XCTAssertEqual(adapter.getCallCount(), 1)

        _ = try await client.mutations.createCachedUser(input: kCreateInput)
        XCTAssertEqual(adapter.getCallCount(), 2)

        _ = try await client.queries.getCachedUser(id: "user-1")
        XCTAssertEqual(adapter.getCallCount(), 3)
    }

    func test_resetAll_invalidatesAllTags_allCachedQueriesRefetch() async throws {
        adapter.registerData("getCachedUser", n(("getCachedUser", Fixtures.kUserAliceJson)))
        adapter.registerData("getCachedPost", n(("getCachedPost", Fixtures.kPostJson)))
        adapter.registerData("resetAll", n(("resetAll", true)))

        _ = try await client.queries.getCachedUser(id: "user-1")
        _ = try await client.queries.getCachedPost(id: "post-1")
        XCTAssertEqual(adapter.getCallCount(), 2)

        _ = try await client.mutations.resetAll()
        XCTAssertEqual(adapter.getCallCount(), 3)

        _ = try await client.queries.getCachedUser(id: "user-1")
        _ = try await client.queries.getCachedPost(id: "post-1")
        XCTAssertEqual(adapter.getCallCount(), 5)
    }
}

// ── Cross-tag isolation ──────────────────────────────────────────────────────

final class CacheTest_CrossTagIsolation: XCTestCase {
    var adapter: MockAdapter!
    var client: GraphLinkClient!

    override func setUp() {
        adapter = MockAdapter()
        client = Fixtures.newClient(adapter, MockWebSocketAdapter())
    }

    func test_usersTagInvalidation_doesNotAffect_postsCache() async throws {
        adapter.registerData("getCachedPost", n(("getCachedPost", Fixtures.kPostJson)))
        adapter.registerData("createCachedUser", n(("createCachedUser", Fixtures.kUserAliceJson)))

        _ = try await client.queries.getCachedPost(id: "post-1")
        XCTAssertEqual(adapter.getCallCount(), 1)

        _ = try await client.mutations.createCachedUser(input: kCreateInput)
        XCTAssertEqual(adapter.getCallCount(), 2)

        _ = try await client.queries.getCachedPost(id: "post-1")
        XCTAssertEqual(adapter.getCallCount(), 2)
    }

    func test_transferPost_invalidatesBothTags() async throws {
        adapter.registerData("getCachedUser", n(("getCachedUser", Fixtures.kUserAliceJson)))
        adapter.registerData("getCachedPost", n(("getCachedPost", Fixtures.kPostJson)))
        adapter.registerData("transferPost", n(("transferPost", Fixtures.kPostJson)))

        _ = try await client.queries.getCachedUser(id: "user-1")
        _ = try await client.queries.getCachedPost(id: "post-1")
        XCTAssertEqual(adapter.getCallCount(), 2)

        _ = try await client.mutations.transferPost(postId: "post-1", newAuthorId: "user-2")
        XCTAssertEqual(adapter.getCallCount(), 3)

        _ = try await client.queries.getCachedUser(id: "user-1")
        _ = try await client.queries.getCachedPost(id: "post-1")
        XCTAssertEqual(adapter.getCallCount(), 5)
    }
}

// ── listCachedUsers ───────────────────────────────────────────────────────────

final class CacheTest_ListCachedUsers: XCTestCase {
    var adapter: MockAdapter!
    var client: GraphLinkClient!

    override func setUp() {
        adapter = MockAdapter()
        client = Fixtures.newClient(adapter, MockWebSocketAdapter())
    }

    func test_listCachedUsers_isCachedAfterFirstCall() async throws {
        adapter.registerData("listCachedUsers", n(("listCachedUsers", [Fixtures.kUserAliceJson])))
        _ = try await client.queries.listCachedUsers()
        _ = try await client.queries.listCachedUsers()
        XCTAssertEqual(adapter.getCallCount(), 1)
    }

    func test_createCachedUser_invalidatesListCachedUsers() async throws {
        adapter.registerData("listCachedUsers", n(("listCachedUsers", [Fixtures.kUserAliceJson])))
        adapter.registerData("createCachedUser", n(("createCachedUser", Fixtures.kUserAliceJson)))

        _ = try await client.queries.listCachedUsers()
        _ = try await client.mutations.createCachedUser(input: kCreateInput)
        _ = try await client.queries.listCachedUsers()
        XCTAssertEqual(adapter.getCallCount(), 3)
    }
}

// ── staleIfOffline ────────────────────────────────────────────────────────────

final class CacheTest_StaleIfOffline: XCTestCase {
    var adapter: MockAdapter!
    var client: GraphLinkClient!

    override func setUp() {
        adapter = MockAdapter()
        client = Fixtures.newClient(adapter, MockWebSocketAdapter())
    }

    func test_staleData_isReturnedWhenNetworkFails_afterTtlExpires() async throws {
        adapter.registerData("getStaleUser", n(("getStaleUser", Fixtures.kUserAliceJson)))

        _ = try await client.queries.getStaleUser(id: "user-1")
        XCTAssertEqual(adapter.getCallCount(), 1)

        // wait for TTL to expire (TTL = 1s, uses real time)
        try await Task.sleep(nanoseconds: 1_100_000_000)

        adapter.simulateFailure = true
        let res = try await client.queries.getStaleUser(id: "user-1")

        XCTAssertNotNil(res.getStaleUser)
        XCTAssertEqual(res.getStaleUser!.id, "user-1")
    }

    func test_nonStaleEntry_throwsWhenNetworkFails() async throws {
        adapter.registerData("getCachedUser", n(("getCachedUser", Fixtures.kUserAliceJson)))
        adapter.simulateFailure = true
        do {
            _ = try await client.queries.getCachedUser(id: "user-1")
            XCTFail("expected throw")
        } catch {
            // expected
        }
    }
}
